Solutions/Corvus.Monitoring.AspnetCore.Mvc.Demo/Controllers/BaseController.cs

```diff
@@ -0,0 +1,28 @@
+﻿// <copyright file="BaseController.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Corvus.Monitoring.AspnetCore.Mvc.Demo.Controllers
+{
+    using Corvus.Monitoring.Instrumentation;
+    using Microsoft.AspNetCore.Mvc;
+
+    /// <summary>
+    /// Base class for controllers in this application.
+    /// </summary>
+    /// <remarks>
+    /// Having a base class for your controllers that implements <see cref="IHaveObservableActionMethods "/>
+    /// and has the <see cref="ObservableActionMethodsAttribute"/> action filter is the simplest way of
+    /// adding observability to your controllers.
+    /// </remarks>
+    [ObservableActionMethods]
+    public abstract class BaseController : Controller, IHaveObservableActionMethods
```

Comment:
I think you may already have explained this to me, but I've forgotten. Why do we need both an interface and an attribute?

I read through the rest of the changes in the PR hoping that it would become apparent to me by the end, but I'm not confident I really understand. I _think_ it's that we need the attribute because that's an action filter, and this is what enables us to wedge our code into the request handling. But that alone doesn't give us a way to make the `IOperationInstance` available to the controller, so we've also had to define this interface, which the filter can look for and use to plug in the value. I think it's an error to apply the attribute, and not implement the interface?

This design seems to fit somewhat awkwardly with nullable reference types. I'm wondering if you explored a couple of alternatives that spring to mind (but which might not be feasible, because I've not thought this through):

1. somehow make the `IOperationInstance` available through scoped DI, so you could just declare it as a ctor argument
2. make it available with an extension method on the `HttpContext` (possibly hanging it off the same `IFeature` mechanism that ASP.NET Core seems to use for everything else that can be accessible through the context)

If the 2nd one is possible, you could drop the `IHaveObservableActionsMethods` interface, modify this `BaseController` to have just a readonly property something like:

```cs
public IOperationInstance CurrentOperation => this.Context.GetCurrentOperationInstance();
```

That way, we're showing how to use a property-based solution, but we're avoiding the unsatisfactory nullable type handling. (We might even be able to define an optional interface that provides a default implementation of `CurrentOperation` so you could use the interface as an optional mix-in instead of a thing you were simply required to implement.)

Reply:
The main piece of the code is the action filter. You can use this entirely on its own, without needing to implement the interface, and you'll still get operations recorded for the action method and result execution.

The interface is an added extra; if your controller implements it, the action filter will populate the `CurrentOperation` property which makes it possible for the controller to add properties and metrics to the operation.

As a result, it's not really an error to apply the action filter and not implement the interface; you may not want to add anything to the operation in your controller action so there would be no need to use the interface.

I'll have a look at alternatives; at first glance, stashing it in the `HttpContext` seems like a decent option. From the nullability perspective, I think I'd then want to make the accessor have a non-nullable return type and make it throw if you attempt to access the current operation when it's not set. That way we will avoid silent failure when you think you've set it up correctly but have messed something up.

Reply:
I've modified the code to store the current `IOperationInstance` in the `HttpContext.Items` collection, and provides extension methods to access this. It's had a few side effects:
- Most obviously, it removes the need for `IHaveObservableActionsMethods`. We can't use a method with a default implementation on the interface because it would need to be able to access `HttpContext` (which in a controller is done via `this.Context.HttpContext`).
- It opens up the opportunity for making the `IOperationInstance` created for result execution, available in the same way, meaning you can access it and add data to it from the view. I can't actually see why you'd want to do this, but I've added it for the sake of consistency anyway, along with an example in the demo project.
- The extension method for retrieving the current operation now throws an exception if the `IOperationInstance` isn't present in the context. On reflection, I think this is better than making the return type for that method nullable - essentially, if you're calling the method and the instance isn't present, this suggests you've made an error at dev time.

Regarding your suggestions - I don't think the Features collection is the right place to put this - essentially it's designed for a well-known set of services which provide data to different bits of the `HttpContext` (see https://learn.microsoft.com/en-us/aspnet/core/fundamentals/request-features?view=aspnetcore-7.0) and allows you to switch out implementations as needed. It's not designed for extending the data made available via `HttpContext`.

Adding it to the DI container and wouldn't work in this case because the controller instance is created prior to the action filter methods being executed.